src/main/java/eu/europeana/metis/sandbox/service/workflow/InternalValidationServiceImpl.java

```diff
@@ -68,19 +71,19 @@ public RecordInfo validate(Record recordToValidate) {
   }
 
   private void generateAnalysis(String datasetId, byte[] recordContent) throws PatternAnalysisException {
-    final Lock lock = datasetIdLocksMap.computeIfAbsent(datasetId, s -> new ReentrantLock());
+    final Lock lock = datasetIdLocksMap.computeIfAbsent(datasetId, s -> lockRegistry.obtain("generateAnalysis_" + datasetId));
     final ExecutionPoint executionPoint;
     try {
       lock.lock();
       LOGGER.debug("Generate analysis: {} lock, Locked", datasetId);
       final LocalDateTime timestamp = datasetIdTimestampMap.computeIfAbsent(datasetId, s -> getLocalDateTime(datasetId));
       //We have to attempt initialization everytime because we don't have an entry point for the start of the step
       executionPoint = patternAnalysisService.initializePatternAnalysisExecution(datasetId, Step.VALIDATE_INTERNAL, timestamp);
+      patternAnalysisService.generateRecordPatternAnalysis(executionPoint, new String(recordContent, StandardCharsets.UTF_8));
```

Comment:
FYI:
This was outside of the lock since it was multi-thread safe.
It was kept there as a performance benefit.
Adding it gives more security I suppose but perhaps some performance degredation.
I'm fine with both so we can keep it here for now and see how it behaves.

Reply:
The comment I wrote disappear :sweat_smile: I am writing it again. While developing I notice that `generateRecordPatternAnalysis` has a race condition in the ```DO UPDATE SET record_occurrences = problem_patterns.dataset_problem_pattern.record_occurrences + ? ``` if one node right after the other attempt to update, we will end up with different amount of occurrences that is why I decided to move it back in.